docs/cookbook/extending-the-ui.md

```diff
@@ -0,0 +1,990 @@
+
+:::note
+As a guide for selecting error type, the Apostrophe `array` and `object` fields use `:error="null"`. This provides clear guidance on which field is in error, without a second error message being displayed for the entire component. You may want to change this if your custom field has a requirement such as `min: 1`.
+:::
+
+```javascript
+<template #body>
+  <div class="custom-input-object">
+    <div class="custom-input-wrapper">
+      <AposSchema :schema="gradeSchema" :trigger-validation="triggerValidation" :utility-rail="false" :generation="generation" v-model="gradeSchemaInput">
+      </AposSchema>
+      <span class="letter-grade">{{ letterGrade() }}</span>
+    </div>
+  </div>
+</template>
+```
+
+The `AposSchema` component does the majority of work constructing and displaying our schema fields, as well as passing data.
+
+In this case, we are wrapping it in some additional markup to allow for customized styling. In making your components, this may not be necessary. To return the letter grade we are adding an additional span after the `AposSchema` with content set by the `letterGrade()` method.
+
+The `AposSchema` component typically takes the `:trigger-validation`, `:utility-rail` and `:generation` props. These are used by Apostrophe for field validation, display, and updating.
+
+To keep track of state within our component, we are passing `gradeSchemaInput` that comes from our `data()` method to `v-model`.
+
+The `AposSchema` component takes the array of schema fields passed to `:schema`, renders their inputs and emits a new object with `value` and `hasErrors` sub-properties when changes occur.
+
+**The schema field JavaScript**
+
+At the top of our script, we import all of the mixins and component files that we will be using in our custom component.
+
+```javascript
+import AposInputMixin from 'apostrophe/modules/@apostrophecms/schema/ui/apos/mixins/AposInputMixin';
+import AposInputWrapper from 'apostrophe/modules/@apostrophecms/schema/ui/apos/components/AposInputWrapper.vue';
+import AposSchema from 'apostrophe/modules/@apostrophecms/schema/ui/apos/components/AposSchema.vue';
+```
+
+In this case, we are bringing the `AposInputMixin` mixin into our script. This provides prop declarations for several of the props used in both the `AposInputWrapper` and `AposSchema` components. 
+
+We are also importing the files for the actual components. While not shown here, button controls can be added by bringing in the `<AposButton>` component from `apostrophe/modules/@apostrophecms/ui/ui/apos/components/AposButton.vue`.
+
+```javascript
+export default {
+  name: 'InputGradeField',
```

Comment:
This component name / filename is good - adding Field as a suffix to these is indeed our convention, because not all components are field types.